Standardize on manual instrumentation intro text (#2852)
open-telemetry/opentelemetry.io: layouts/shortcodes/docs/instrumentation/manual-intro.md

Reviewer: Hmmm. I like where this is going, but I think we should clear things up a bit, not helped at all by opentelemetry's confusingly overloaded terminology around "API" and "SDK"

For application developers:

- They install their language's SDK
- They use the SDK to do setup, register dependencies, and critically, exporting the data
- They use the API to instrument
- It's...kind of an uninteresting detail that there's a distinction between API and SDK here, because the SDK they install already pulls in what they need

For library developers:

- They install only the appropriate "API" package for their library
- They instrument using the API
- They _do not_ configure export of any data

I think we need to be very clear about this distinction, which I don't think is clear enough in the current form.

Author: I agree with you, I tried to use as much existing wording as possible, but maybe it is time to rewrite. This intro should be short and refer to the concept page on manual instrumentation for more details, let me see how I can balance that.

Author: I copied some text from the concept page on manual instrumentation, I hope that makes things clearer?

[Manual instrumentation](/docs/concepts/instrumentation/manual/) is the process
of adding observability code to your application yourself. To do so, you need to
import OpenTelemetry to your code. If you're developing a library or some other
component that is intended to be consumed by a runnable binary, then you would

> only take a dependency on the API, because unless the OpenTelemetry SDK, the

Reviewer 2: > I, because unless the OpenTelemetry SDK

The sentense feels incomplete.

Author: apologies, as said in my other comment it looks like I wrote this being not really working properly. Took @cartermp's proposal!

> [Manual instrumentation](/docs/concepts/instrumentation/manual/) is the process
> of adding observability code to your application yourself. To do so, you need to
> import OpenTelemetry to your code. If you're developing a library or some other
> component that is intended to be consumed by a runnable binary, then you would
> only take a dependency on the API, because unless the OpenTelemetry SDK, the
> instrumentation does nothing and does not impact software performance. For more
> details, on instrumenting libraries read
> [these guidelines](/docs/concepts/instrumentation/libraries/).
> 
> If your artifact is a standalone service, then you would take a dependency on
> the API and the SDK.
> 
> For more information about the OpenTelemetry API and SDK, see the
> [specification](/docs/specs/otel/).

Reviewer: I..ended up basically rewriting this. Couldn't help myself. Feel free to take it however you like, though.

```suggestion
[Manual instrumentation](/docs/concepts/instrumentation/manual/) is the act
of adding observability code to your application yourself.

If you're instrumenting an application, you need to use the OpenTelemetry SDK
for your language. You'll then use the SDK to initialize OpenTelemetry and the
API to instrument your code. This will emit telemetry from your app and any
library you installed that also comes with instrumentation.

If you're instrumenting a library, you must only install the OpenTelemetry API
package for your language. This means that your library will only emit telemetry
when an application that uses the OpenTelemetry SDK uses it. Otherwise, your
library will just behave as before. For more on instrumenting libraries, read
[these guidelines](/docs/concepts/instrumentation/libraries/).

For more information about the OpenTelemetry API and SDK, see the
[specification](/docs/specs/otel/).
```

Author: Looks like my brain is malfunctioning to write this down properly 😆 ... I take yours!

[Manual instrumentation](/docs/concepts/instrumentation/manual/) is the act

> of adding observability code to your application yourself.

Reviewer 3: ```suggestion
of adding observability code to an application yourself.
```

> If you're instrumenting an application, you need to use the OpenTelemetry SDK

Reviewer 2: instrumenting part just requires API only.

Reviewer: True, but...it serves no purpose without an exporter, which you need the SDK for.

Author: @reyang what part is not necessarily true? That it serves no purpose without an exporter or that you need the SDK, can you elaborate?

for your language. You'll then use the SDK to initialize OpenTelemetry and the

> API to instrument your code. This will emit telemetry from your app and any

Reviewer 3: I'm ok with your use of "app" here, but let's be consistent and replace "application" in the rest of this diff hunk by "app".

Author: fixed

> API to instrument your code. This will emit telemetry from your app and any

Reviewer 3: ```suggestion
API to instrument your code. This will emit telemetry from your app, and any
```

library you installed that also comes with instrumentation.

> If you're instrumenting a library, you must only install the OpenTelemetry API
> package for your language. This means that your library will only emit telemetry
> when an application that uses the OpenTelemetry SDK uses it. Otherwise, your
> library will just behave as before. For more on instrumenting libraries, read
> [these guidelines](/docs/concepts/instrumentation/libraries/).

Reviewer 3: ```suggestion
If you're instrumenting a library, only install the OpenTelemetry API package
for your language. Your library will not emit telemetry on its own. It will only
emit telemetry when it is part of an application that uses the OpenTelemetry
SDK. For more on instrumenting libraries, see
[Libraries](/docs/concepts/instrumentation/libraries/).
```

> 

Reviewer 2: I'd also recommend including a sentence here about "manual instrumentation does not necessarily mean creating the span/metrics by yourself. One may leverage instrumentation libraries (add link to it)"..

Reviewer: I would disagree with that. At the risk of making this a bikeshedding moment, I think instrumentation libraries are much more in the realm of automatic instrumentation than manual instrumentation. In many cases it's literally just one line of code and then you get a bunch of data.

Reviewer 2: We all have different opinions, but its already clarified here:
https://opentelemetry.io/docs/concepts/instrumentation/automatic/

Reviewer: Well, the definition of manual instrumentation we have here is pretty clear: https://opentelemetry.io/docs/specs/otel/glossary/#manual-instrumentation

And given packages like node's autoinstrumentations-node package, I still stand by the statement that an instrumentation library is more close to automatic instrumentation than manual instrumentation.

Author: This is an [endless](https://github.com/open-telemetry/opentelemetry.io/issues/1689) [discussion](https://github.com/open-telemetry/opentelemetry-specification/issues/2866) and end-user's (aka readers of the documentation) should not be bothered with.

I agree with what @cijothomas said in essence: if an end user is doing manual instrumentation they will make use of 3rd party libraries and they don't want to instrument them themselves, so they either hope that this library is natively instrumented (yeah!) OR they need to make use of an instrumentation library.

If this is no "manual instrumentation" or not, does not really matter for the end user, what matters is that they can and should make use of those to get a better instrumentation experience.

So, I hope we can solve this by wording, e.g.: 

> if you **do** manually instrument your code, you don't have to create all telemetry by yourself: your dependencies either provide native instrumentation for you, or you can make use of instrumentation libraries. To learn more about instrumentation libraries, read [this]. 

My goal is to choose a wording that neither implies the one or the other stance. 

P.S.: If you like to continue the bikesheeding, I refer to the existing discussions on this topic I linked above and, for `autoinstrumentation-node` package additionally, see https://github.com/open-telemetry/opentelemetry-js/issues/3298 (although things changed because this package now actually contains the "auto instrumentation": https://opentelemetry.io/docs/instrumentation/js/automatic/

Reviewer: I don't think that's a good paragraph to add/modify. In our own definition we make it very clear that manual instrumentation means using the API to instrument yourself.

The confusion around "automatic" that other maintainers have self-inflicted is just rearing its head again here, because we have components that do not require the use of the API to instrument, and yet they're somehow not considered "automatic instrumentation" or something close to that. It's just weird. You either get components that create data for you or you have to create that data yourself.

Author: > I don't think that's a good paragraph to add/modify. In our own definition we make it very clear that manual instrumentation means using the API to instrument yourself.

I actually think the opposite is true and adding this sentence is valuable to end-users, since this question comes up lately often like "can I do manual instrumentation but still use instrumentation library X for my dependency".

Again, that's why I suggested we phrase it in a way that does not suggest that Instrumentation Libraries are manual instrumentation, they can supplement it (if this is the right English word) or you can combine manual instrumentation and instrumentation libraries like you can combine manual instrumentation with the javaagent in java's case (and the same is true for the equivalent in most other languages (actively avoiding the A* word here))

If we don't find consensus here, let's create a follow-up PR to discuss that and try to get this one closed.

Reviewer: That bring us into the bikeshed about agents, though 🙂 

I won't die on this hill, for the sake of this PR, but I fear deeply that we're inching towards a path of "technically correct in the way the maintainers reason about things" rather than "what end-users typically care about". Perhaps I'm wrong and people typically see adding an instrumentation library as some form of manual instrumentation, in which case I would propose that we change that definition instead.

Author: > I won't die on this hill, for the sake of this PR,

Me neither, that's why I propose we merge this PR for now and create a follow-up one for the other discussion. If this requires a change of definition, then let's see if we re-vive that battle on the spec 😆

[Manual instrumentation](/docs/concepts/instrumentation/manual/)

> is the process of adding observability code to your application yourself.

Reviewer 3: ```suggestion
is the process of adding observability code to an application yourself.
```

> To do so, you need to import OpenTelemetry to your code. If you're

Reviewer 3: ```suggestion
To do so, your app code needs to import OpenTelemetry. If you're
```

> developing a library or some other component that is intended to be consumed by

Reviewer 3: ```suggestion
developing a library or component that will be part of
```

Author: outdated

> a runnable binary, then you would only take a dependency on the API, because

Reviewer 3: ```suggestion
a runnable binary, then your could would only depend on the API, because
```

> unless the OpenTelemetry SDK, the instrumentation does nothing and does not

Reviewer 3: ```suggestion
without the OpenTelemetry SDK, the instrumentation code has no effect and so does not 
```

impact software performance. For more details, on instrumenting libraries read

> [these guidelines](/docs/concepts/instrumentation/libraries/).

Reviewer 3: ```suggestion
impact software performance. For more details on instrumenting libraries, see 
[Libraries](/docs/concepts/instrumentation/libraries/).
```

Author: outdated

> If your artifact is a standalone service, then you would take a dependency on

Reviewer 3: ```suggestion
If your component is a standalone service, then it will depend on both
```
